### PR TITLE
[Compiled Autograd] Inplace updates of gradients

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106176
* #106172
* #105887
* #105809
* #105814
* #105813
* #105808
* __->__ #105713

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/46168

**Comment by jansel** [comment]
> Related? #46168

I don't think so.  This is about reproducing existing eager behavior in compiled autograd.

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 1]
```diff
@@ -65,32 +65,35 @@ void AccumulateGrad::compiled_args(CompiledNodeArgs& args) {
...
   at::Tensor grad_copy = variable.grad();
   saved.before(variable_copy);
   saved.before(grad_copy);
-  int callback_count = 0;
+
+  // TODO(jansel): accumulateGrad() below will only do the `+=`-style
+  // update when GradMode is disabled.  GradMode seems to get disabled
+  // in eager mode, but somewhere else in the stack.  Need to figure
+  // out where that happens, doing it here is wrong.
+  at::NoGradGuard guard;
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/csrc/autograd/functions/accumulate_grad.cpp:86]: Thanks!

File: torch/csrc/dynamo/compiled_autograd.h [hunk 2]
```diff
@@ -94,19 +94,69 @@ struct NodeCalls : public std::unordered_map<Node*, NodeCall> {
...
+  bool defined() const {
+    return id != 0;
+  }
+  uint32_t id;
+  at::Tensor proxy_tensor;
+};
+
+struct TensorArgs {
+  // Manages a collection of TensorArgs and mappings from Tensors/SavedVariables
+  // to them
```
Comment (albanD) [inline torch/csrc/dynamo/compiled_autograd.h:113]: Should we mention that the whole point is to be able to unpack() saved_var exactly one and keep a mapping to the unpacked Tensor?

File: torch/csrc/dynamo/python_compiled_autograd.cpp [hunk 3]
```diff
@@ -449,20 +461,17 @@ variable_list compiled_autograd(
...
-    TORCH_INTERNAL_ASSERT(
-        outputs.size() == compiler_call.set_grad_targets.size());
+    TORCH_INTERNAL_ASSERT(outputs.size() == cache->output_grad_targets.size());
     for (const auto i : c10::irange(outputs.size())) {
-      // TODO(jansel): does this one need to be an inplace copy?  if so it
-      // should go in the graph
-      at::Tensor& grad = compiler_call.set_grad_targets[i].mutable_grad();
-      grad = outputs[i];
+      compiler_call.tensor_args.inputs[cache->output_grad_targets[i]]
+          .mutable_grad() = outputs[i];
```
Comment (albanD) [inline torch/csrc/dynamo/python_compiled_autograd.cpp:474]: In the inplace case, we have `compiler_call.tensor_args.inputs[cache->output_grad_targets[i]].mutable_grad().arg.unsafeGetTensorImpl() == outputs[i].arg.unsafeGetTensorImpl()` leading to the grad field being preserved.
Can we mention that in the comment above?

**Review by albanD — COMMENTED** [review]
Very nice to see the removal of the need for exact order matching!

File: torch/csrc/dynamo/python_compiled_autograd.cpp (hunk 3 above)
Comment (jansel) [reply torch/csrc/dynamo/python_compiled_autograd.cpp:474]: In the inplace case the callback lambda is never called, and this code isn't run.  The graph just contains a `add_` op in it on one of its inputs (and returns nothing).

Notice how there is no call to `update_grad()` in this case:
https://github.com/pytorch/pytorch/blob/4f76be8c77acd933c580512cdb955a6cd6e95a4a/torch/csrc/autograd/functions/accumulate_grad.h#L209

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 4]
```diff
@@ -64,32 +64,28 @@ void AccumulateGrad::compiled_args(CompiledNodeArgs& args) {
...
-  int callback_count = 0;
   accumulateGrad(
       variable_copy,
       grad_copy,
       grads[0],
-      0 /* num_expected_refs, 0 disables in-place update */,
-      [&callback_count, &saved](const at::Tensor& grad_update) {
-        callback_count++;
-        saved.set_grad_value(grad_update);
+      0 /* num_expected_refs, 0 disables aliased reuse */,
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/csrc/autograd/functions/accumulate_grad.cpp:85]: Yeah, the compiler should be able to remove the extra clone.

File: torch/csrc/dynamo/python_compiled_autograd.cpp (hunk 3 above)
Comment (albanD) [reply torch/csrc/dynamo/python_compiled_autograd.cpp:474]: Ho right.
Could we have some comment that explains how accumulate grad is handled and the different cases of what might happen?

Comment (jansel) [reply torch/csrc/dynamo/python_compiled_autograd.cpp:474]: Done

**Review by albanD — APPROVED** [review]
Thanks for the update!

**Pull request closed without merging**